src/main/java/net/sf/mzmine/modules/peaklistmethods/alignment/adap3/ADAP3AlignerTask.java

```diff
@@ -0,0 +1,419 @@
+
+    private final MZmineProject project;
+    private final ParameterSet parameters;
+
+    private final PeakList[] peakLists;
+
+    private final String peakListName;
+
+    private final Project alignment;
+
+    public ADAP3AlignerTask(MZmineProject project, ParameterSet parameters) {
+
+        this.project = project;
+        this.parameters = parameters;
+
+        this.peakLists = parameters.getParameter(
+                ADAP3AlignerParameters.PEAK_LISTS)
+                .getValue().getMatchingPeakLists();
+
+        this.peakListName = parameters.getParameter(
+                ADAP3AlignerParameters.NEW_PEAK_LIST_NAME).getValue();
+
+        this.alignment = new Project();
+    }
+
+    @Override
+    public String getTaskDescription() {
+        return "ADAP Aligner, " + peakListName + " (" + peakLists.length
+                + " peak lists)";
+    }
+
+    @Override
+    public double getFinishedPercentage() {
+        return alignment.getProcessedPercent();
+    }
+
+    @Override
+    public void run() {
+
+        if (!isCanceled()) {
+            String errorMsg = null;
+
+            setStatus(TaskStatus.PROCESSING);
+            LOG.info("Started ADAP Peak Alignment");
+
+            try {
+                PeakList peakList = alignPeaks(peakLists);
+
+                if (!isCanceled()) {
+                    project.addPeakList(peakList);
+
+                    QualityParameters.calculateQualityParameters(peakList);
+
+                    setStatus(TaskStatus.FINISHED);
+                    LOG.info("Finished ADAP Peak Alignment");
+                }
+            } catch (IllegalArgumentException e) {
+                errorMsg = "Incorrect Peak Lists:\n" + e.getMessage();
+            } catch (Exception e) {
+                errorMsg = "'Unknown error' during alignment. \n"
+                        + e.getMessage();
+            } catch (Throwable t) {
+                setStatus(TaskStatus.ERROR);
+                setErrorMessage(t.getMessage());
+                LOG.log(Level.SEVERE, "ADAP Alignment error", t);
+            }
+
+            // Report error
+            if (errorMsg != null) {
+                setErrorMessage(errorMsg);
+                setStatus(TaskStatus.ERROR);
+            }
+        }
+    }
+
+    private PeakList alignPeaks(PeakList[] peakLists) {
+
+        // Collect all data files
+
+        List<RawDataFile> allDataFiles = new ArrayList<>(peakLists.length);
+
+        for (final PeakList peakList : peakLists) {
+            RawDataFile[] dataFiles = peakList.getRawDataFiles();
+            if (dataFiles.length != 1)
+                throw new IllegalArgumentException("Found more then one data "
+                        + "file in some of the peaks lists");
+
+            allDataFiles.add(dataFiles[0]);
+        }
+
+        // Perform alignment
+
+        for (final PeakList peakList : peakLists) {
+
+            Sample sample = new Sample(peakList.hashCode());
```

Comment:
hashCode() is not necessarily unique. I think you should just make a Hashtable Integer -> PeakList for an increasing Integer counter.